fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/ImplementNestedLoopJoinRule.java

```diff
@@ -144,7 +151,7 @@ public void onMatch(@Nonnull PlannerRuleCall call) {
         // as creating a join with that outer and inner combination would leave no place for the middle to be planned.
         //
 
-        final var innerDependencies = fullCorrelationOrder.get(outerAlias);
+        final var innerDependencies = fullCorrelationOrder.get(innerAlias);
 
         for (final var quantifier : selectExpression.getQuantifiers()) {
```

Comment:
Why is this check done here? isn't possible to push it to the matching logic itself such that we never get an invalid outer/inner pair w.r.t. correlation

Reply:
I am assuming you mean _matching_ as in matching the pre of the rule not as in _index matching_. I suppose that's possible in a way that you code every check for each rule into a `BindingMatcher` that is specific for that rule. If you pushed the this check into a `BindingMatcher` you could maybe reuse the matcher in `ImplementExistentialNestedLoopJoin`. I think I am more in the camp of structural vs semantic checks. This can be done as a semantic check as part of the rule instead of as a `BindingMatcher`. It's also using a correlation order (which is memoized and thus cheap to get over and over again) so it felt more natural to do it like this.